csrc/scheduler/normalization_inner_outer.cpp

```diff
@@ -199,31 +440,108 @@ std::pair<int64_t, int64_t> getPersistentBufferSize(
+          outer_broadcast_tvs.empty());
 
-  // in combined_inner_outer_reduction, the partial results of outer
-  // reductions must be persistent, allow register spill avoid segmentation
-  std::vector<TensorView*> outer_reduction_tvs;
-  for (auto tv : reduction_tvs) {
-    if (!scheduler_utils::isFastestDimReduction(tv)) {
-      outer_reduction_tvs.emplace_back(tv);
+  auto total_buffer_size = buffer_params.project_to_input
+      ? persistent_buffer_size_info.projected_persistent_buffer_size
+      : persistent_buffer_size_info.persistent_buffer_size;
+  total_buffer_size +=
+      partialOuterReductionBufferSize(reduction_tvs, runtime_info);
+
+  const auto dev_prop = at::cuda::getCurrentDeviceProperties();
+  int64_t smem_overhead = scheduler_utils::getSharedMemoryOverheadPerBlock(
+      fusion,
+      reduction_tvs,
+      InnerOuterPersistentKernelScheduler::threads_per_block_max);
+  int64_t available_smem =
+      (int64_t)dev_prop->sharedMemPerMultiprocessor - smem_overhead;
+  int64_t available_regs = scheduler_utils::register_file_size_full;
+  buffer_params.smem_overhead = smem_overhead;
+
+  // Put all the persistent tensors in registers
+  buffer_params.regs_buffer_size = total_buffer_size;
+  buffer_params.smem_buffer_size = 0;
+
+  // Relocate buffers to shared memory until the buffer size in registers is
+  // within the allowable limit.
+  if (buffer_params.regs_buffer_size > available_regs) {
+    const auto sorted_candidate_tvs = buffer_params.project_to_input
+        ? sortProjectableBufferInputs(
+              fusion,
+              persistent_buffer_info.persistent_buffers,
+              persistent_buffer_info.projectable_buffer_inputs,
+              outer_broadcast_tvs,
+              reduction_tvs,
+              runtime_info)
+        : sortPersistentBuffers(persistent_buffer_info.persistent_buffers);
+
+    // determine the number of buffers to move to shared memory
+    const int64_t n_buffers = (int64_t)sorted_candidate_tvs.size();
+    int64_t acc_regs_buffer_size = 0, acc_smem_buffer_size = 0;
+    int64_t n_smem_buffer = -1;
+    for (int i = 1; i <= n_buffers; i++) {
+      auto current_tv = sorted_candidate_tvs[i - 1];
+      int64_t tv_buffer_size_regs =
+          scheduler_utils::getPersistentBufferSizeOfTensor(
+              current_tv, runtime_info, persistent_buffer_info);
+
+      int64_t tv_buffer_size_smem = roundUpSharedMemory(
+          tv_buffer_size_regs,
+          dataTypeSize(current_tv->getDataType().value()),
+          vectorize_factor,
+          InnerOuterPersistentKernelScheduler::threads_per_block_min,
+          InnerOuterPersistentKernelScheduler::threads_per_block_max,
+          dev_prop->warpSize);
+
+      acc_regs_buffer_size += tv_buffer_size_regs;
+      acc_smem_buffer_size += tv_buffer_size_smem;
+      // check if we have enough registers after moving the first-i buffers to
+      // shared memory
+      if (buffer_params.regs_buffer_size - acc_regs_buffer_size <=
```

Comment:
Unclear what's being done here. `regs_buffer_size` is the total buffer size (btw, shouldn't it be renamed to to just `total_buffer_size` as it doesn't necessarily indicate the register usage). `acc_regs_buffer_size` is the sum of the register sizes of all the tensors visited so far. `regs_buffer_size - acc_regs_buffer_size` seems to indicate the necessary shmem size if the first i buffers are moved to shared memory. Am I following the code correctly? If so, what does `<= available_regs` mean?

Reply:
Now it's much more clear. Thanks!